python/cuml/internals/input_utils.py

```diff
@@ -0,0 +1,516 @@
+        pass
+
+    if hasattr(X, "__cuda_array_interface__"):
+        return cp.ndarray
+
+    if hasattr(X, "__array_interface__"):
+        # For some reason, numpy scalar types also implement
+        # `__array_interface__`. See numpy.generic.__doc__. Exclude those types
+        # as well as np.dtypes
+        if (not isinstance(X, np.generic) and not isinstance(X, type)):
+            return np.ndarray
+
+    try:
+        if cupyx.scipy.sparse.isspmatrix(X):
+            return cupyx.scipy.sparse.spmatrix
+    except UnavailableError:
+        pass
+
+    try:
+        if (scipy_sparse.isspmatrix(X)):
+            return scipy_sparse.spmatrix
+    except UnavailableError:
+        pass
+
+    # Return None if this type isnt supported
+    return None
+
+
+def determine_array_type(X):
+    if (X is None):
+        return None
+
+    # Get the generic type
+    gen_type = get_supported_input_type(X)
+
+    return _input_type_to_str.get(gen_type, None)
+
+
+def determine_array_dtype(X):
+
+    if (X is None):
+        return None
+
+    if isinstance(X, (CudfDataFrame, PandasDataFrame)):
+        # Assume single-label target
+        dtype = X[X.columns[0]].dtype
+    else:
+        try:
+            dtype = X.dtype
+        except AttributeError:
+            dtype = None
+
+    return dtype
+
+
+def determine_array_memtype(X):
+    try:
+        return X.mem_type
+    except AttributeError:
+        pass
+    if hasattr(X, '__cuda_array_interface__'):
+        return MemoryType.device
+    if hasattr(X, '__array_interface__'):
+        return MemoryType.host
+    if isinstance(X, (CudfDataFrame, CudfSeries)):
+        return MemoryType.device
+    if isinstance(X, (PandasDataFrame, PandasSeries)):
+        return MemoryType.device
+    return None
+
+
+def determine_array_type_full(X):
+    """
+    Returns a tuple of the array type, and a boolean if it is sparse
+
+    Parameters
+    ----------
+    X : array-like
+        Input array to test
+
+    Returns
+    -------
+    (string, bool) Returns a tuple of the array type string and a boolean if it
+        is a sparse array.
+    """
+    if (X is None):
+        return None, None
+
+    # Get the generic type
+    gen_type = get_supported_input_type(X)
+
+    if (gen_type is None):
+        return None, None
+
+    return _input_type_to_str[gen_type], gen_type in _sparse_types
+
+
+def is_array_like(X):
+    try:
+        return (
+            hasattr(X, '__cuda_array_interface__')
+            or (
+                hasattr(X, '__array_interface__')
+                and not (
+                    isinstance(X, global_settings.xpy.generic)
+                    or isinstance(X, type)
+                )
+            ) or isinstance(X, (
+                SparseCumlArray, CudfSeries, PandasSeries, CudfDataFrame,
+                PandasDataFrame))
+            or numba_cuda.devicearray.is_cuda_ndarray(X)
+        )
+    except UnavailableError:
+        return False
```

Comment:
Maybe better to just raise? In what scenario would it be sensible to continue in this case?

Reply:
The logic of this function is all messed up for the "unavailable" case. I've updated accordingly. In general, we want to return False to indicate that it's _not_ array-like when we get `UnavailableError`, since we can be confident that it's not the type we attempted to check. We can take another look at this with the corrected logic.